### PR TITLE
Refactor data access repositories for modular pipeline change

### LINKED ISSUE #1899: Refactor Modular Pipelines in the Kedro-viz backend
## Description

The logic of how we calculate modular pipelines in Kedro-viz backend differs from how it is computed in Kedro. For instance, modular pipelines are not associated with namespaces, and neither are the datasets in a modular pipeline. Only nodes have namespaces, and based on the way the datasets are consumed by different nodes, these datasets can be regarded as input/output/children to each modular pipeline.

This PR is to revise the current logic and make it more like the way Kedro does it. Modular Pipelines have caused a lot of issues for the user based on the difference in approach. With this refactor, we hope to align the approaches.

## Context
Why is this change important to you? How would you use it? How can it benefit other users?

## Possible Implementation
(Optional) Suggest an idea for implementing the addition or change.

## Possible Alternatives
(Optional) Describe any alternative solutions or features you've considered.

## Checklist

- [ ] Include labels so that we can categorise your feature request

### PR DESCRIPTION
**Pull request by ravi-kumar-pilla**
## Description

Partially resolves #1899 

## Development notes

* Create a dictionary of node and modular pipeline mapping for faster search
* Add parameters set to keep track of parameters (As parameter nodes do not have modular pipelines)
* Add `populate_tree` method which takes care of inputs/outputs and children for all the modular pipelines within a registered pipeline
* Add `explode_namespace` method which takes care of expanding a nested namespace.
* Modify add_inputs and add_outputs methods to take a `Set[str]` of inputs and outputs that are calculated in `populate_tree` method
* Add `add_children` method which resolves the modular pipeline child nodes
* Add `_add_children_to_parent_pipeline` method which resolves adding children to parent modular pipeline in-case of nesting
* Add `_add_datasets_as_children` method which adds datasets as children to the modular pipeline
* Update tests

## QA notes

- This PR is part of a bigger refactor (https://github.com/kedro-org/kedro-viz/pull/1897) of modular pipelines and is created to ease review process.
- The CI build might fail as this PR is not self-sufficient

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 1]
```diff
@@ -1,246 +1,324 @@
...
+            List[str]: A list of expanded namespaces.
+            Example:
+                >>> nested_namespace = "uk.data_processing.internal"
+                >>> modular_pipeline_repo_obj = ModularPipelinesRepository()
+                >>> expanded_namespace =
+                        modular_pipeline_repo_obj._explode_namespace(nested_namespace)
+                >>> expanded_namespace
+                ["uk", "uk.data_processing", "uk.data_processing.internal"]
+        """
+        if not nested_namespace or "." not in nested_namespace:
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:98]: Why is it necessary to check for `if not nested_namespace`? And then again below: `[nested_namespace] if nested_namespace else []`?

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 2]
```diff
@@ -1,246 +1,324 @@
...
+        ns_parts = nested_namespace.split(".")
+        return [".".join(ns_parts[: i + 1]) for i in range(len(ns_parts))]
 
     def get_or_create_modular_pipeline(
         self, modular_pipeline_id: str
     ) -> ModularPipelineNode:
-        """Get the modular pipeline node with the given ID from the repository.
-        If it doesn't exist, create the node, add to the repository and return the instance.
+        """
+        Retrieve or create a modular pipeline node.
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:107]: Why change the doc string? The behaviour is still the same right? The previous one also seems to provide more explanation.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 3]
```diff
@@ -1,246 +1,324 @@
...
-        input_node: Union[DataNode, TranscodedDataNode, ParametersNode],
-    ) -> None:
-        """Add an input to a modular pipeline based on whether it's an internal or external input.
-        The input to a modular pipeline can only be a data node or parameter node.
-        The input also has knowledge of which modular pipelines it belongs to
-        based on its namespace. This information can be accessed with through
-        `input_node.modular_pipelines`.
+    def add_inputs(self, modular_pipeline_id: str, inputs: Set[str]) -> None:
+        """
+        Add input datasets to the modular pipeline.
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:124]: In the code there's also a special case for `parameters`. I think it makes sense to mention that in the docstring.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 4]
```diff
@@ -1,246 +1,324 @@
...
+        """
+        hashed_outputs = {self._hash_input_output(output) for output in outputs}
+        self.tree[modular_pipeline_id].outputs = hashed_outputs
 
-        is_internal_output = modular_pipeline_id in output_node.modular_pipelines
-        if is_internal_output:
-            self.tree[modular_pipeline_id].internal_outputs.add(output_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _hash_input_output(self, item: str) -> str:
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:151]: Why does this need to take `self`? As far as I understand this method just hashes input/output so it doesn't need an instance of `ModularPipelinesRepository`

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 5]
```diff
@@ -1,246 +1,324 @@
...
+            free_inputs = sub_pipeline.inputs()
+            free_outputs = sub_pipeline.outputs() | (
+                rest_of_the_pipeline.inputs() & sub_pipeline.all_outputs()
+            )
+
+            self.add_inputs(modular_pipeline_id, free_inputs)
+            self.add_outputs(modular_pipeline_id, free_outputs)
+            self.add_children(modular_pipeline_id, sub_pipeline.nodes)
+
+    def _explode_namespace(self, nested_namespace: str) -> List[str]:
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:81]: Why does this need to take `self`? This doesn't need an instance of `ModularPipelinesRepository` so it can just be a static method or it can be defined outside the `ModularPipelinesRepository` class altogether.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 6]
```diff
@@ -1,246 +1,324 @@
...
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _hash_input_output(self, item: str) -> str:
+        """Hash the input/output dataset."""
+        return (
+            _hash(_strip_transcoding(item))
+            if TRANSCODING_SEPARATOR in item
+            else _hash(item)
+        )
+
+    def add_children(self, modular_pipeline_id: str, kedro_nodes: List[Node]):
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:159]: Why are this method and `add_outputs` and `add_inputs` public?

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 7]
```diff
@@ -1,246 +1,324 @@
...
+                and io_id not in self.parameters
+            ):
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=io_id, type=GraphNodeType.DATA)
+                )
+                if io_id not in self.node_mod_pipeline_map:
+                    self.node_mod_pipeline_map[io_id] = set()
+                self.node_mod_pipeline_map[io_id].add(modular_pipeline.id)
+
+    def has_modular_pipeline(self, modular_pipeline_id: str) -> bool:
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:309]: I'm not sure this method is necessary, it's only called in one place as far as I can tell and IMO actually makes it more effort to figure out what's going on.

**Review by merelcht — COMMENTED** [review]
I've done an initial review of this part, but want to have another look after reviewing the other parts. I've left comments mostly related to readability of the code.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 8]
```diff
@@ -1,246 +1,324 @@
...
-    TranscodedDataNode,
 )
+from kedro_viz.utils import TRANSCODING_SEPARATOR, _strip_transcoding
+
+
+def _hash(value: str):
+    return hashlib.sha1(value.encode("UTF-8")).hexdigest()[:8]
+
+
+def is_dataset_param(dataset_name: str) -> bool:
```
[the remark this answers is not in the document]
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:24]: Not really. I observed this method is used via Catalog Repository as well. I am shifting this to utils.py which will help in removing code duplication. Thank you

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 5 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:81]: Yes, converted to static method. Thank you

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 1 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:98]: I think we should only check for the below condition -

```python
if not nested_namespace:
    return []
```
Thanks

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 2 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:107]: Yes, it is the same. Reverting back this change.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 3 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:124]: Yes there is a special case for parameters and that is while adding the datasets as children. That is in the docstring for `add_children`, `_add_children_to_parent_pipeline` and `_add_datasets_as_children`. The method here is to add_inputs to the modular pipeline (like the free_inputs hanging outside the modular pipeline)

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 4 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:151]: Yes, I would like to know where should the helper functions like this be placed in ? I used to place all the helper functions in utils as public. But there should be some rules which I am not aware of. Could you please let me know ? Thanks

Placed in -

1. A class as static method (public/private)
2. Outside of a class but in same file (public/private)
3. Inside utils (public/private)

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 6 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:159]: I do not see them being used outside this class. They were public before the code change as they were used in managers.py. But now it is not needed. Thank you for pointing it out

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 4 above)
Comment (merelcht) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py]: This stackoverflow thread, specifically the first two answers, describe quite well what the general rule of thumb is when it comes to static methods, class methods and module functions. https://stackoverflow.com/questions/11788195/module-function-vs-staticmethod-vs-classmethod-vs-no-decorators-which-idiom-is

In the Kedro codebase we generally apply the following:
1. Inside a class as static method: any method that is only used by that class, but the difference with a module function is that these types of methods are specific to this class and don't have any meaning outside of it. (e.g. https://github.com/kedro-org/kedro/blob/main/kedro/io/data_catalog.py#L336-L338)
2. Outside of a class but in same file: any helper method that is only used in that specific file should be a module function at the top of the file. Note that these methods are only used in that file/class, but don't require any "knowledge" about the class: no need to access a class instance and the method has meaning outside of the context of the class (e.g. https://github.com/kedro-org/kedro/blob/main/kedro/framework/session/session.py#L33-L56) 
3. Inside utils: when it's a helper method used in multiple files and not specific to a certain class. 

When it comes to making things public/private it's a bit different for Viz vs. Kedro but in the framework anything that's _public_ is part of the public API. So it can (and will) be used by outside users directly accessing Kedro as a library. So that means, unless a user should have access to the method, we make it _private_. My personal opinion is that for Viz we should make things _public_ when it's accessed in multiple classes/files.

Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py]: Thank you for the explanation Merel 💯

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 9]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
-        """
+                modular_pipeline.tags.update(kedro_node.tags)
 
-        # There is no need to extract modular pipeline from parameters
-        # because all valid modular pipelines are encoded in either a TaskNode or DataNode.
-        if isinstance(node, ParametersNode):
-            return None
+                # add input/output datasets of a kedro node as a child
+                # to the modular pipeline
+                self._add_datasets_as_children(
```
Comment (rashidakanchwala) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:204]: I was thinking we could add datasets that have no modular pipeline ID associated with them (and thus belong to the root modular pipeline) in this function instead of in managers.py.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 10]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+                self.node_mod_pipeline_map[dataset] = set()
+            self.node_mod_pipeline_map[dataset].add(modular_pipeline_id)
+
+    def _add_datasets_as_children(
+        self,
+        modular_pipeline: ModularPipelineNode,
+        kedro_node: Node,
+        modular_pipeline_inputs_outputs: Set[str],
+    ):
+        """Helper to add datasets (not parameters) related to task nodes as children.
```
Comment (rashidakanchwala) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:295]: I was thinking we could add datasets that have no modular pipeline ID associated with them (and thus belong to the root modular pipeline) in this function instead of in managers.py.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 11]
```diff
@@ -1,79 +1,124 @@
...
+    return hashlib.sha1(value.encode("UTF-8")).hexdigest()[:8]
+
+
+def _hash_input_output(item: str) -> str:
+    """Hash the input/output dataset."""
+    return (
+        _hash(_strip_transcoding(item))
+        if TRANSCODING_SEPARATOR in item
+        else _hash(item)
+    )
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:30]: I would put these in `utils` and import them here, rather than the transcoding logic. Hide as much as possible behind a smaller API footprint.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 12]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
-            )
+        hashed_outputs = {_hash_input_output(output) for output in outputs}
+        self.tree[modular_pipeline_id].outputs = hashed_outputs
 
-        is_internal_output = modular_pipeline_id in output_node.modular_pipelines
-        if is_internal_output:
-            self.tree[modular_pipeline_id].internal_outputs.add(output_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _add_children(self, modular_pipeline_id: str, kedro_nodes: List[Node]):
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:168]: This whole function is overly complicated, very wide open if condition and too much nestedness. Could you try to refactor it into a more concise and easier to follow one?

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 13]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
-    def extract_from_node(self, node: GraphNode) -> Optional[str]:
-        """Extract the namespace from a graph node and add it as a modular pipeline node
-        to the modular pipeline repository.
+        for kedro_node in kedro_nodes:
+            # add kedro node as a child to the modular pipeline
+            if kedro_node.namespace == modular_pipeline_id:
+                kedro_node_id = _hash(str(kedro_node))
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=kedro_node_id, type=GraphNodeType.TASK)
+                )
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:198]: Instead of nesting this with 2 extra levels, you could do something like:

```python
children = [n for n in kedro_nodes if n.namespace == modular_pipeline_id]
modular_pipeline.children = [ModularPipelineChild(id=_hash(str(n)), type=GraphNodeType.TASK) for n in children]
```

Much more readable and explainable:
1. We get all the nodes matching exactly the pipeline namespace
2. Add them as children to the pipeline (this one could be `.extend` instead of `=` if there's other places that add to children called before this one)

Two actions, two lines of code...

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 14]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+                    modular_pipeline, kedro_node, modular_pipeline_inputs_outputs
+                )
 
-        modular_pipeline_id = node.namespace
-        if not modular_pipeline_id:
-            return None
+                # update node_modular_pipeline lookup
+                if kedro_node_id not in self.node_mod_pipeline_map:
+                    self.node_mod_pipeline_map[kedro_node_id] = set()
+                self.node_mod_pipeline_map[kedro_node_id].add(modular_pipeline_id)
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:211]: If `self.node_mod_pipeline_map` was a `defaultdict(set)`, these 3 lines can be simplified to:

```
self.node_mod_pipeline_map[kedro_node_id].add(modular_pipeline_id)
```

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 15]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+            parent_modular_pipeline.tags.update(modular_pipeline.tags)
 
-        modular_pipeline.tags.update(node.tags)
+            # add current modular pipeline and input/output datasets
+            # of a modular pipeline as a child to the parent modular
+            # pipeline based on the rules
+            self._add_children_to_parent_pipeline(
+                parent_modular_pipeline,
+                modular_pipeline_id,
+                modular_pipeline_inputs_outputs,
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:235]: This is very unintuitive - the method is called `_add_children`, but in it we have the totally unrelated code for adding the modular pipeline itself to it's parent's children. Could this be factored out as a separate method which is called right after `_add_children` instead?

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 16]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+                dataset not in parent_node.inputs
+                and dataset not in parent_node.outputs
+                and dataset not in self.parameters
+            ):
+                parent_node.children.add(
+                    ModularPipelineChild(id=dataset, type=GraphNodeType.DATA)
+                )
+            if dataset not in self.node_mod_pipeline_map:
+                self.node_mod_pipeline_map[dataset] = set()
+            self.node_mod_pipeline_map[dataset].add(modular_pipeline_id)
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:287]: `defaultdict` to the rescue here...

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 17]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+            if (
+                io_id not in modular_pipeline_inputs_outputs
+                and io_id not in self.parameters
+            ):
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=io_id, type=GraphNodeType.DATA)
+                )
+                if io_id not in self.node_mod_pipeline_map:
+                    self.node_mod_pipeline_map[io_id] = set()
+                self.node_mod_pipeline_map[io_id].add(modular_pipeline.id)
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:316]: `defaultdict`...

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 18]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+        """
+        hashed_io_ids = {
+            _hash_input_output(io)
+            for io in set(kedro_node.inputs) | set(kedro_node.outputs)
+        }
+        for io_id in hashed_io_ids:
+            if (
+                io_id not in modular_pipeline_inputs_outputs
+                and io_id not in self.parameters
+            ):
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:310]: This filter could've just been added to the previous `for`-comprehension, or as a separate one on its own and then we don't need the extra nestedness.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 19]
```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
...
+                )
+                if io_id not in self.node_mod_pipeline_map:
+                    self.node_mod_pipeline_map[io_id] = set()
+                self.node_mod_pipeline_map[io_id].add(modular_pipeline.id)
+
+    def get_node_and_modular_pipeline_mapping(
+        self, node
+    ) -> Tuple[str, Union[Set[str], None]]:
+        """Get the modular pipeline(s) to which the given node belongs."""
+        node_id = (
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:322]: To avoid things like this, I would've created a bit more clever `_hash` helper, which can correctly hash the dataset or the node.

**Review by idanov — COMMENTED** [review]
From a correctness point of view, this looks good as it is as far as I can tell. There's a few things that can be done to make the style of the code a bit more readable. I've added some comments here and there with pointers. Happy to discuss on a call if you have questions.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 19 above)
Comment (rashidakanchwala) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:322]: dataset or task node*   -- i tried to make this function a bit better but in this function a  'node' can be both dataset or task node. But u r right, the use of 'node' is getting a bit confusing.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 20]
```diff
@@ -1,79 +1,124 @@
...
+def _hash_input_output(item: str) -> str:
+    """Hash the input/output dataset."""
+    return (
+        _hash(_strip_transcoding(item))
+        if TRANSCODING_SEPARATOR in item
+        else _hash(item)
+    )
 
 
 class ModularPipelinesRepository:
```
[the remark this answers is not in the document]
Comment (rashidakanchwala) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:33]: This is a problem throughout the codebase. I think we need to clarify definitions here and create another refactor PR that tackles this bit.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 19 above)
Comment (idanov) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:301]: I also like the reversal here, start with the shorter one and finish with the longer one.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 21]
```diff
@@ -165,162 +162,146 @@ def _add_outputs(self, modular_pipeline_id: str, outputs: Set[str]) -> None:
...
-            parent_node (ModularPipelineNode): The parent modular pipeline node.
+            modular_pipeline (ModularPipelineNode): The modular pipeline node.
             modular_pipeline_id (str): The ID of the modular pipeline to add as a child.
-            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs
-                    to/from the modular pipeline
+            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs to/from
+            the modular pipeline.
         """
+        # Determine the parent modular pipeline ID
+        parent_modular_pipeline_id = (
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:265]: I'd simply call this `parent_id`, variables have context and in this function the only thing that can be a parent is a modular pipeline, thus it's implied and doesn't need to be part of the variable name.

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 22]
```diff
@@ -165,162 +162,146 @@ def _add_outputs(self, modular_pipeline_id: str, outputs: Set[str]) -> None:
...
-            # add current modular pipeline and input/output datasets
-            # of a top-level modular pipeline as a child to ROOT modular
-            # pipeline
-            self._add_children_to_parent_pipeline(
-                self.tree[ROOT_MODULAR_PIPELINE_ID],
-                modular_pipeline_id,
-                modular_pipeline_inputs_outputs,
+            modular_pipeline.tags.update(node.tags)
+
+            hashed_io_ids = {
```
Comment (idanov) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:222]: I would name this just `io_ids`.

**Review by idanov — APPROVED** [review]

File: package/kedro_viz/data_access/repositories/modular_pipelines.py (hunk 3 above)
Comment (merelcht) [reply package/kedro_viz/data_access/repositories/modular_pipelines.py:138]: I still think it's a good idea to describe the following code in the docstring:

```
 if is_dataset_param(_input):
                self.parameters.add(hashed_input)
```

File: package/kedro_viz/data_access/repositories/modular_pipelines.py [hunk 23]
```diff
@@ -83,163 +125,185 @@ def get_or_create_modular_pipeline(
...
-        )
-        return modular_pipeline_id
+    def _add_children_to_parent_pipeline(
+        self,
+        modular_pipeline: ModularPipelineNode,
+        modular_pipeline_id: str,
+        modular_pipeline_inputs_outputs: Set[str],
+    ):
+        """
+        Helper to add modular_pipeline children correctly to parent modular pipelines
```
Comment (merelcht) [inline package/kedro_viz/data_access/repositories/modular_pipelines.py:248]:
```suggestion
        Helper to add modular pipeline children correctly to parent modular pipelines

```

**Review by merelcht — APPROVED** [review]
Left some final minor comments, but nothing blocking!
